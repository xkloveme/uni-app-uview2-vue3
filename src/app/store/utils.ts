export { Store, Pinia } from 'pinia-class-component'
import { createDecorator } from 'pinia-class-component'

/**
 *  持久化装饰器
 */
export const Persist = createDecorator(
  (options, key) =>
    (options['persist'] ||= {
      enabled: true,
      strategies: [
        {
          key: 'pinia-persist-' + options.name,
          storage: {
            getItem: key => uni.getStorageSync(key),
            setItem: (key, data) => uni.setStorage({ key, data }),
            removeItem: key => uni.removeStorage({ key }),
          },
          paths: [],
        },
      ],
    }) && options['persist'].strategies[0].paths.push(key),
)
