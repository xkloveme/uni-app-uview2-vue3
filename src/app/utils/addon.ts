Promise.prototype.wait = function (ms) {
  return this.then(
    res => new Promise(resolve => setTimeout(() => resolve(res), ms)),
    err => new Promise((_, reject) => setTimeout(() => reject(err), ms)),
  )
}

Promise.prototype.go = function () {
  return this.then(
    res => new Promise(resolve => resolve([res, null])),
    err => new Promise(resolve => resolve([null, err])),
  )
}

Promise.wait = ms => new Promise(r => setTimeout(r, ms))
Promise.do = async fn => await fn()

interface Promise<T> {
  wait(ms: number): Promise<T>
  go(): Promise<[T?, any?]>
}
interface PromiseConstructor {
  wait(n: number): Promise<void>
  do<T extends (...args: any[]) => any>(fn: T): Promise<Awaited<ReturnType<T>>>
}

// 小程序Global补全
// #ifdef MP
Object.assign(global, {
  Array: Array,
  Date: Date,
  Error: Error,
  Function: Function,
  Math: Math,
  Object: Object,
  RegExp: RegExp,
  String: String,
  TypeError: TypeError,
  setTimeout: setTimeout,
  clearTimeout: clearTimeout,
  setInterval: setInterval,
  clearInterval: clearInterval,
})
// #endif
