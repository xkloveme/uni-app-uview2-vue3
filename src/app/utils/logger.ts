let f: any = ({ title, args }) => console.log(` [ ${title} ] `, ...args)

// h5和微信平台有devools
// #ifdef H5 | MP
f = ({ type, bg, title, args }) =>
  console[['log', 'info', 'warn', 'error'].includes(type) ? type : 'log'](
    `%c ${title} `,
    `color:#fff;background:${bg};padding:5px;margin:5px 0;font-family:Monaco,PingFang SC`,
    ...args,
  )
// #endif

// app平台调试环境有devools
// #ifdef APP-PLUS
try {
  if (global)
    f = ({ type, bg, title, args }) =>
      console[['log', 'info', 'warn', 'error'].includes(type) ? type : 'log'](
        `%c ${title} `,
        `color:#fff;background:${bg};padding:5px;margin:5px 0;font-family:Monaco,PingFang SC`,
        ...args,
      )
} catch (error) {}
// #endif

const types = {
  log: 'linear-gradient(45deg, #4facfe 0%, #00f2fe 150%);',
  info: 'linear-gradient(120deg, #a1c4fd 0%, #c2e9fb 100%);',
  warn: 'linear-gradient(to right, #e6b980 0%, #eacda3 100%);',
  error: 'linear-gradient(45deg, #ff0844 0%, #ffb199 100%);',
  debug: 'linear-gradient(45deg, #ff9a9e 0%, #fad0c4 99%, #fad0c4 100%)',
  success: 'linear-gradient(45deg, #43e97b 0%, #38f9d7 150%);',
}

export const loggers = fromEntries(
  Object.keys(types).map(type => {
    let [bg] = types[type].split('|')
    // f({ type, bg, title: 'test ' + type, args: [] }) // test
    return [type, (title, ...args) => f({ type, bg, title, args })]
  }),
) as { [K in keyof typeof types]: (title: string, ...args: any[]) => void }

export default function () {
  Object.assign(app, loggers)
}

declare global {
  interface App extends Is<typeof loggers> {}
}

function fromEntries(iterable) {
  return [...iterable].reduce((obj, [key, val]) => {
    obj[key] = val
    return obj
  }, {})
}
