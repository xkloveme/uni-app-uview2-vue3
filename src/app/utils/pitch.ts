import { App as VueApp } from 'vue'

let patch = {
  log: <T>(...args: [T, ...any]) => (
    app.debug(/src(.*?\.vue)/.exec(new Error('123').stack)?.[1] || '页面输出', ...args), args[0]
  ),
  any: arg => arg as any,
}

export default function (vueApp: VueApp) {
  Object.assign(vueApp.config.globalProperties, patch)
}

type patchType = typeof patch

declare module '@vue/runtime-core' {
  export interface ComponentCustomProperties extends patchType {}
}
