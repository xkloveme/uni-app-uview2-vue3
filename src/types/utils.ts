export interface PageParams {
  data?: any
  id?: string
  from?: string
}
