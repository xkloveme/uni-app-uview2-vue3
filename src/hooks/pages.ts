import { ToRefs } from 'vue'
import { PageParams } from '@/types'

type QueryDefaultData<T> = Omit<PageParams, 'data'> & { data: T & AObjectHasAnyKeys }
/**
 * 获得当前页面跳转携带的参数和信息
 * (依赖 app.to )
 * @param fn 函数时为回调函数,其他为默认返回值
 */
export function useQuery<T>(fn: T): ToRefs<QueryDefaultData<T>>
export function useQuery(fn?: (data: any) => void): ToRefs<PageParams>
export function useQuery<T>(fn?: (data: any) => void | T) {
  let isF = typeof fn === 'function'
  const query = reactive<{ data?: any; id?: string; from?: string }>({
    data: isF ? {} : fn,
    id: null,
    from: null,
  })

  Promise.do(async () => {
    if (!getCurrentPages().pop()?.['$page']?.fullPath) await new Promise(r => onLoad(r))
    let id = getCurrentPages().pop()['$page'].fullPath.split('?id=')[1]
    uni.$emit(id + '_query', pkg => Object.assign(query, pkg))
    isF && fn(query.data)
  }).catch(err => app.error('参数获取失败', err))

  return toRefs(query)
}

import { Mescroll } from '@/types'
export const ScrollSymbol = Symbol('mescroll')
export interface ScrollOptions {
  mescroll?: Mescroll
  enable: 'none' | 'up' | 'down' | 'all'
  fetch: (...arg: any[]) => void
}
/**
 * 上拉加载,下拉刷新
 *
 * @example
 * useScroll(onPageScroll)
 *   .onLoad((page) => {
 *     page.endSuccess()
 *   })
 */
export function useScroll(onPageScroll?: typeof import('@dcloudio/uni-app')['onPageScroll']) {
  const scrollOptions = reactive<ScrollOptions>({
    enable: 'all',
    mescroll: null,
    fetch: page => setTimeout(() => page.endSuccess(10, false), 1000),
  })
  provide(ScrollSymbol, scrollOptions)
  let { mescroll, fetch, enable } = $(scrollOptions)
  onPageScroll?.(e => mescroll && mescroll.onPageScroll(e))
  onReachBottom(() => mescroll && mescroll.onReachBottom())
  onPullDownRefresh(() => mescroll && mescroll.onPullDownRefresh())

  const o = {
    onLoad: (cb: (mescroll: Mescroll) => void) => ((fetch = cb), o),
    enable: (type: ScrollOptions['enable']) => ((enable = type), o),
    mescroll: $$(mescroll),
  }
  return o
}
