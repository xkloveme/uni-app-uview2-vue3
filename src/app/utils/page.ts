import { debounce } from 'lodash'
import { encode } from 'js-base64'
import { useQuery } from '@/hooks'

const pageTools = {
  to: debounce(
    (path: string, obj?: any) => {
      const currentPath = [...getCurrentPages()].pop()['route'].split('/')
      const t = String(+new Date())
      if (!path.startsWith('/')) {
        let pathArray = path.split('/')
        path = [
          null,
          ['', currentPath[0], currentPath[1], pathArray[0]],
          ['', currentPath[0], pathArray[0], pathArray[1]],
        ][pathArray.length].join('/')
      }
      const pkg = { data: obj, from: currentPath, t }
      //TODO: 改成事件机制
      const querystr = encode(JSON.stringify(pkg)).replace(/=/g, '等')
      uni.navigateTo({
        url: path + '?q=' + querystr,
        fail: err => console.log(err),
        success: () => app.info(`页面切换`, obj || '', '=>' + path),
      })
      return new Promise<any>(r => uni.$once(t, r))
    },
    500,
    { leading: true, trailing: false },
  ),
  back: debounce(
    (data?: any) => {
      if (data) {
        console.log('回调:', data)
        const { t } = useQuery()
        uni.$emit(t.value, data)
      }

      //TODO: 这里有bug,不能用try来判断成功与否
      try {
        uni.navigateBack({})
      } catch (error) {
        const url = '/pages/index/index'
        try {
          uni.navigateTo({ url })
        } catch (error) {
          uni.switchTab({ url })
        }
      }
    },
    500,
    { leading: true, trailing: false },
  ),
}

export default function () {
  Object.assign(app, pageTools)
}

type pageType = typeof pageTools

declare global {
  interface App extends pageType {}
}
