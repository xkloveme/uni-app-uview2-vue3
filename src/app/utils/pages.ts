import { debounce, DebouncedFuncLeading } from 'lodash'
import { useQuery } from '@/hooks'

export function getPath(p: string, currentGroup: string) {
  let [group, path] = Array.from(p.match(/^(?:#(.*?)(?:\/|$))?(.*)$/)).slice(1)
  if (group === undefined) return path
  else return `/pages/${group || currentGroup}/${path || 'index'}`
}

let to = debounce(
  (path: string, obj?: any) => {
    const currentPath = getCurrentPages().pop()['route']
    const currentGroup = currentPath.split('/')[1]
    const id = uni.$u.guid(5) + new Date().getTime()

    const pkg = { data: obj, from: currentPath, id }
    uni.$on(id + '_query', cb => cb(pkg))
    let url = getPath(path, currentGroup)
    uni
      .navigateTo({ url: url + '?id=' + id })
      .then(() => app.info('页面切换', obj || '', '=>', path == url ? path : `${path} (${url})`))
      .catch(err => app.error('页面切换失败', '=>' + path, err))
    return new Promise<any>((r, e) => (uni.$once(id + '_resolve', r), uni.$once(id + '_reject', e)))
  },
  50,
  { leading: true, trailing: false },
)

const back = debounce(
  (data?: any, type: 'resolve' | 'reject' = 'resolve') => {
    if (data) console.log('回调:', data)
    const { id } = useQuery()
    uni.$emit(id.value + '_' + type, data)
    const url = '/pages/index/index'
    if (getCurrentPages().length > 1) uni.navigateBack()
    else uni.redirectTo({ url }).catch(() => uni.switchTab({ url }))
  },
  50,
  { leading: true, trailing: false },
) as DebouncedFuncLeading<(data?: any, type?: 'resolve' | 'reject') => void> & {
  resolve: (data?: any) => void
  reject: (data?: any) => void
}

back.resolve = (data?: any) => back(data, 'resolve')
back.reject = (data?: any) => back(data, 'reject')

const pageTools = {
  to,
  back,
}

export default function () {
  Object.assign(app, pageTools)
}

type pageType = typeof pageTools

declare global {
  interface App extends pageType {}
}
