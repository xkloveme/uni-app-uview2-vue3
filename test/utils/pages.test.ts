import { hashKeyAsGroupAliasTargets } from './pages.targets'
import { getPath } from '@/app/utils/pages'
describe('页面工具', () => {
  test('使用#作为包名标识', async () => {
    const pathOutput = hashKeyAsGroupAliasTargets.map(
      path => (path + ' ').padEnd(20, '-') + '-> ' + getPath(path, '$currentGroup'),
    )
    expect(pathOutput).toMatchSnapshot()
  })
})
