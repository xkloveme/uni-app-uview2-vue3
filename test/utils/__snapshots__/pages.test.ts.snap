// Vitest Snapshot v1

exports[`页面工具 > 使用#作为包名标识 1`] = `
[
  "#group --------------> /pages/group/index",
  "#group/ -------------> /pages/group/index",
  "#group/foo ----------> /pages/group/foo",
  "#/foo/bar -----------> /pages/$currentGroup/foo/bar",
  "# -------------------> /pages/$currentGroup/index",
  "#/ ------------------> /pages/$currentGroup/index",
  "foo/bar -------------> foo/bar",
  "foo/bar/baz ---------> foo/bar/baz",
  "/pages/foo/bar ------> /pages/foo/bar",
  "./foo/bar -----------> ./foo/bar",
  "../foo --------------> ../foo",
]
`;
